src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java

```diff
@@ -106,6 +113,79 @@ public final long pendingSidelineMessagesCount() {
+     * @param unit : timeout unit
+     * @return : List of message nacked
+     * @throws Exception
+     */
+    public final List<Message> publishWithConfirmListener(List<Message> messages, AMQP.BasicProperties properties,
+            long timeout, @NotNull TimeUnit unit) throws Exception {
+        publishChannel.confirmSelect();
+        ConcurrentNavigableMap<Long, Message> outstandingConfirms = new ConcurrentSkipListMap<>();
+        final CountDownLatch publishAckLatch = new CountDownLatch(messages.size());
+        publishChannel.addConfirmListener((sequenceNumber, multiple) -> {
+            if (multiple) {
+                ConcurrentNavigableMap<Long, Message> confirmed = outstandingConfirms.headMap(
+                        sequenceNumber, true
+                );
+                for(int i =0;i<confirmed.size();i++)
+                    publishAckLatch.countDown();
+                confirmed.clear();
+            } else {
+                publishAckLatch.countDown();
+                outstandingConfirms.remove(sequenceNumber);
+            }
+            log.info(String.format("Message with delivery tag %s acknoledged", sequenceNumber));
+        }, (sequenceNumber, multiple) -> {
+            {
+                if(multiple == true)
+                {
+                    ConcurrentNavigableMap<Long, Message> confirmed = outstandingConfirms.headMap(
+                            sequenceNumber, true
+                    );
+                    for(int i =0;i<confirmed.size();i++)
+                        publishAckLatch.countDown();
+                }
+                else
+                    publishAckLatch.countDown();
+            }
+            log.debug(String.format("Message nacked with multiple= %s : %s", multiple, outstandingConfirms.get(sequenceNumber)));
+        });
+
+        String routingKey;
+        if (config.isSharded()) {
+            routingKey = NamingUtils.getShardedQueueName(queueName, getShardId());
+        } else {
+            routingKey = queueName;
+        }
+
+        long startTime = System.nanoTime();
+        for (Message message : messages) {
+            try {
+                outstandingConfirms.put(publishChannel.getNextPublishSeqNo(), message);
+                publishChannel.basicPublish(config.getExchange(), routingKey, properties,
+                        mapper().writeValueAsBytes(message));
+            } catch (Exception e) {
+                log.debug(String.format("Failed to publish Message : %s with exception %s", message, e));
+                publishAckLatch.countDown();
+            }
+        }
+
+        if (!publishAckLatch.await(unit.toMillis(timeout), TimeUnit.MILLISECONDS)) {
+            log.error("Timed out waiting for publish acks");
+        }
+
+        long endTime = System.nanoTime();
+        log.info(String.format("Published %,d messages with confirmListener in %,d ms%n", messages.size() - outstandingConfirms.size(),
```

Comment:
Could you change the message to also contain the number of total messages? Maybe something like
("Published %,d/%,d messages with confirmListener in %d ms", publishedCount, totalCount, duration)
I don't think the last %n is required.

Reply:
changed